crates/flipperzero/src/storage.rs

```diff
@@ -0,0 +1,252 @@
+        }
+    }
+}
+
+/// File stream with buffered read operations.
+pub struct File(*mut sys::File);
+
+impl File {
+    pub fn new() -> Self {
+        unsafe {
+            File(sys::storage_file_alloc(
+                UnsafeRecord::open(RECORD_STORAGE).as_ptr(),
+            ))
+        }
+    }
+}
+
+impl Drop for File {
+    fn drop(&mut self) {
+        unsafe {
+            // `storage_file_close` calls `storage_file_sync`
+            // internally, so it's not necesssary to call it here.
+            sys::storage_file_close(self.0);
+        }
+    }
+}
+
+impl Read for File {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
+        let to_read = buf.len().try_into().map_err(|_| Error::InvalidParameter)?;
+        let bytes_read =
+            unsafe { sys::storage_file_read(self.0, buf.as_mut_ptr() as *mut c_void, to_read) };
+        if bytes_read == to_read || bytes_read < to_read && unsafe { sys::storage_file_eof(self.0) }
+        {
+            Ok(bytes_read as usize)
+        } else {
+            Err(Error::from_sys(unsafe {
+                sys::storage_file_get_error(self.0)
+            }))
+        }
+    }
+}
+
+impl Seek for File {
+    fn seek(&mut self, pos: SeekFrom) -> Result<usize, Error> {
+        let (offset_type, offset) = match pos {
+            SeekFrom::Start(n) => (true, n.try_into().map_err(|_| Error::InvalidParameter)?),
+            SeekFrom::End(n) => (false, n.try_into().map_err(|_| Error::InvalidParameter)?),
+            SeekFrom::Current(n) => (false, n.try_into().map_err(|_| Error::InvalidParameter)?),
+        };
+        unsafe {
+            if sys::storage_file_seek(self.0, offset, offset_type) {
+                Ok(sys::storage_file_tell(self.0)
+                    .try_into()
+                    .map_err(|_| Error::InvalidParameter)?)
+            } else {
+                Err(Error::from_sys(sys::storage_file_get_error(self.0)))
+            }
+        }
+    }
+
+    fn rewind(&mut self) -> Result<(), Error> {
+        self.seek(SeekFrom::Start(0)).map(|_| {})
+    }
+
+    fn stream_len(&mut self) -> Result<usize, Error> {
+        Ok(unsafe {
+            sys::storage_file_size(self.0)
+                .try_into()
+                .map_err(|_| Error::InvalidParameter)?
+        })
+    }
+
+    fn stream_position(&mut self) -> Result<usize, Error> {
+        Ok(unsafe {
+            sys::storage_file_tell(self.0)
+                .try_into()
+                .map_err(|_| Error::InvalidParameter)?
+        })
+    }
+}
+
+impl Write for File {
+    fn write(&mut self, buf: &[u8]) -> Result<usize, Error> {
+        let to_write = buf.len().try_into().map_err(|_| Error::InvalidParameter)?;
+        let bytes_written =
+            unsafe { sys::storage_file_write(self.0, buf.as_ptr() as *mut c_void, to_write) };
+        if bytes_written == to_write
+            || bytes_written < to_write && unsafe { sys::storage_file_eof(self.0) }
```

Comment:
Same kind of comment here as for `Read::read`: is `storage_file_write` guaranteed to write as many bytes as it can, or only `n <= buf.len()`?